### ISSUE TITLE
Selenium: Running multiple browser parallel using Undetected Chromedriver

### ISSUE THREAD

**Issue by Jonsch-HTL** [seleniumbase/SeleniumBase#1694]
Im currently trying to run multible test-automation at the same time, using multithreading, but for some reason it just runs the automation for the last opened browser. The other opened browser just remain in: page not found. As my Webdriver I am using Undetected Chromedriver aswell as I am using a set UserAgent and a chrome-profile, which i get by accessing the --user-data-dir--.
Does anyone have an idea on how to get the all automations running.

**Comment by mdmintz** [comment]
You can use the ``seleniumbase`` / ``pytest`` built-in multi-threading feature. Here's an example:

Run the following file with ``python`` or ``pytest --uc -s -n=3``

```python
from parameterized import parameterized
from seleniumbase import BaseCase

if __name__ == "__main__":
    from pytest import main
    main([__file__, "--uc", "-s", "-n=3"])

class UndetectedTest(BaseCase):
    @parameterized.expand([[]] * 3)
    def test_browser_is_undetected(self):
        self.open("https://nowsecure.nl/#relax")
        try:
            self.assert_text("OH YEAH, you passed!", "h1", timeout=7.25)
            self.post_message("Selenium wasn't detected!", duration=2.8)
            self._print("\n Success! Website did not detect Selenium! ")
        except Exception:
            self.fail('Selenium was detected! Try using: "pytest --uc"')
```

Output:

```
python test_verify.py 
============================================ test session starts ============================================
platform darwin -- Python 3.11.1, pytest-7.2.1, pluggy-1.0.0
configfile: pytest.ini
plugins: html-2.0.1, forked-1.4.0, rerunfailures-11.0, xdist-3.1.0, ordering-0.6, cov-4.0.0, metadata-2.0.4, seleniumbase-4.11.8
gw0 [3] / gw1 [3] / gw2 [3]

 Success! Website did not detect Selenium! 

 Success! Website did not detect Selenium! 
..
 Success! Website did not detect Selenium! 
.
============================================= 3 passed in 8.85s
```

As for the ``user_data_dir`` option, there's an existing Chromium bug with setting the ``user_data_dir`` on Chrome 109. Titus Fortner from the Selenium Org already opened a ticket with them: https://bugs.chromium.org/p/chromedriver/issues/detail?id=4328

On top of that, there were already existing problems with setting a ``user_data_dir`` with the main ``undetected-chromedriver`` repo: https://github.com/ultrafunkamsterdam/undetected-chromedriver/search?q=user-data-dir&type=issues (See all the open issues)

``seleniumbase`` does fix several issues with ``undetected-chromedriver``, as well as improving on abilities to bypass detection, but this ``user_data_dir`` issue is rooted deep with them, and probably not fixable until Chromium ships a fix to address Titus's open ticket.

As for running multiple browsers in undetected mode simultaneously with SeleniumBase, that's easy to do, and people do that all the time lately (it seems). See the example at the top.

**Comment by Jonsch-HTL** [comment]
First of all thanks for your fast reply. 
I've tried the code above using another url and as it seems it just opens the multiple browser one by one and not at the same time. Did I do something wrong or is that the only way to do so?

**Comment by mdmintz** [comment]
@Jonsch-HTL Run that exact script as is with ``pytest --uc -s -n=3`` (make sure to use ALL the args provided).

That script creates three separate copies of the same test via ``@parameterized.expand([[]] * 3)``. Then the ``-n=3`` arg tells pytest to run tests using multiple processes.

**Comment by phamxuanphongson** [comment]
> You can use the `seleniumbase` / `pytest` built-in multi-threading feature. Here's an example:
> 
> Run the following file with `python` or `pytest --uc -s -n=3`
> 
> ```python
> from parameterized import parameterized
> from seleniumbase import BaseCase
> 
> if __name__ == "__main__":
>     from pytest import main
>     main([__file__, "--uc", "-s", "-n=3"])
> 
> class UndetectedTest(BaseCase):
>     @parameterized.expand([[]] * 3)
>     def test_browser_is_undetected(self):
>         self.open("https://nowsecure.nl/#relax")
>         try:
>             self.assert_text("OH YEAH, you passed!", "h1", timeout=7.25)
>             self.post_message("Selenium wasn't detected!", duration=2.8)
>             self._print("\n Success! Website did not detect Selenium! ")
>         except Exception:
>             self.fail('Selenium was detected! Try using: "pytest --uc"')
> ```
> 
> Output:
> 
> ```
> python test_verify.py 
> ============================================ test session starts ============================================
> platform darwin -- Python 3.11.1, pytest-7.2.1, pluggy-1.0.0
> configfile: pytest.ini
> plugins: html-2.0.1, forked-1.4.0, rerunfailures-11.0, xdist-3.1.0, ordering-0.6, cov-4.0.0, metadata-2.0.4, seleniumbase-4.11.8
> gw0 [3] / gw1 [3] / gw2 [3]
> 
>  Success! Website did not detect Selenium! 
> 
>  Success! Website did not detect Selenium! 
> ..
>  Success! Website did not detect Selenium! 
> .
> ============================================= 3 passed in 8.85s
> ```
> 
> As for the `user_data_dir` option, there's an existing Chromium bug with setting the `user_data_dir` on Chrome 109. Titus Fortner from the Selenium Org already opened a ticket with them: https://bugs.chromium.org/p/chromedriver/issues/detail?id=4328
> 
> On top of that, there were already existing problems with setting a `user_data_dir` with the main `undetected-chromedriver` repo: https://github.com/ultrafunkamsterdam/undetected-chromedriver/search?q=user-data-dir&type=issues (See all the open issues)
> 
> `seleniumbase` does fix several issues with `undetected-chromedriver`, as well as improving on abilities to bypass detection, but this `user_data_dir` issue is rooted deep with them, and probably not fixable until Chromium ships a fix to address Titus's open ticket.
> 
> As for running multiple browsers in undetected mode simultaneously with SeleniumBase, that's easy to do, and people do that all the time lately (it seems). See the example at the top.

Thank you so much for explaining, but I want do it a little bit different. I don't want to use:

**Comment by phamxuanphongson** [comment]
@mdmintz Thank you so much for explaining, but how to still do the same stuffs but without this:
from pytest import main
   main([__file__, "--uc", "-s", "-n=3"])

Because I don't run it from __main__

**Comment by mdmintz** [comment]
@phamxuanphongson That part is there so that if you run the script with ``python``, it still calls ``pytest`` with the correct args that you need. But you have to use ``pytest`` for the multi-process functionality to work, as that uses https://github.com/pytest-dev/pytest-xdist, which is what makes the multi-threading parts work correctly. That was the key to making multi-threaded UC Mode tests work correctly.

**Comment by Jonsch-HTL** [comment]
When i run the code, pytest runs the test one time, and when finished starts the second run, ec.

**Comment by mdmintz** [comment]
It's working for me. What OS? SeleniumBase version? Run command? And exact script used?

**Comment by Jonsch-HTL** [comment]
I'm on Windows 11, SeleniumBase is on newest version and i just copy pasted ur script

**Comment by mdmintz** [comment]
Yep, I see problems now when trying to make it work on Windows. I'm investigating...

**Comment by mdmintz** [comment]
OK, if I use a virtual environment on Windows, it is working for me. The browsers will launch one-at-a-time, but then they run in parallel.

**Comment by mdmintz** [comment]
Fixed better in `4.17.12` - https://github.com/seleniumbase/SeleniumBase/issues/2037
